distributed/scheduler.py

```diff
@@ -2765,7 +2951,71 @@ def bulk_schedule_after_adding_worker(self, ws: WorkerState):
         ordering, so the recommendations are sorted by priority order here.
         """
         ts: TaskState
-        tasks = []
+        recommendations: dict[str, str] = {}
+
+        # Redistribute the tasks between all worker queues. We bubble tasks off the back of the most-queued
```

Comment:
Agreed that there's probably some repetition. I'm not sure how much that could be factored out. This is a pretty different approach from work stealing AFAIU.

Without this though, I think new workers would be mostly/entirely idle. Since work-stealing doesn't know about `queued` tasks, and co-assignment means the candidates for downstream dependencies will only be existing workers, I think it's quite likely new workers wouldn't get used at all. That didn't feel acceptable to me.

Anyway, I wrote this just to profile and experiment with the rebalancing approach on actual clusters. Maintaining good co-assignment while scaling up is an interesting problem (that we don't currently handle). It's not necessarily what would actually get merged.

But I do think we'd need something to rebalance queued tasks during scaling _as long as we're pre-assigning all tasks to workers upfront_. This is the main reason why I'd like to not do that anymore and make co-assignment logic  stateless. It makes handling scaling way, way easier.